Added serialization and deserialization methods for AutogradTensor

DanyEle opened the pull request:
In response to issues raised in #2318 , adding methods to serialize and deserialize an AutogradTensor.

mari-linhares commented on test/torch/tensors/test_native.py:
```diff
@@ -117,15 +117,15 @@ def test_copy():
 
 
-def test_size():
+def test_local_size():
```
why were these tests renamed?

DanyEle replied:
This is just some leftover from #2343

mari-linhares replied:
Would you mind undoing the changes in this file if possible? so we can keep the PR as simple and narrow as possible

DanyEle replied:
Yeah, I'm going to do that tomorrow morning for sure

mari-linhares commented on syft/frameworks/torch/tensors/interpreters/autograd.py:
```diff
@@ -265,3 +265,60 @@ def float_precision(self):
         if isinstance(self.child, torch.Tensor) and not self.child.is_wrapper:
             return self.child
         return self
+
```
Can you add tests for this?

It would be great to have simple tests for serialize / deserialize and also a test where you:

- train a model (with some toy data)
- run inference in the model (with another batch of toy data)
- serialize and deserialize the model
- check that the evaluation has the same result as the original model
- continue training the model and check that the weights are changing

DanyEle replied:
Thanks for the comments. I tried to perform training of an RNN using Autogradtensors just yesterday, and I bumped into some issues related to backpropagation yielding errors, because tensors were wrapped into AutogradTensors. I fixed this issue by extracting the tensors via an "hotfix" into the Pytorch's library code at `torch/autograd/__init__.py` in the `def backward` method.

At line 87, we just had:

```python grad_tensors = list(grad_tensors)```

I replaced that line  with:

```python
 try:
            grad_tensors = list(grad_tensors)
        except:
            print("Gradients extracted!")
            #extract the tensor from the autogradTensor in every tensor
            grad_tensors = extract_grads_from_autograd_tensors(grad_tensors)

def extract_grads_from_autograd_tensors(auto_grad_tensors):
    grad_tensors = []
    for i in range(len(auto_grad_tensors)):
        #if the child was not yet extracted, then do extract it
        
        tensor_grad = auto_grad_tensors[i].child
        grad_tensors.append(tensor_grad)
        
    return(grad_tensors)
```










Any suggestions on how this hotfix could be introduced into PySyft for autograd_training with Autogradtensors?

mari-linhares replied:
What's the exception you get when you try to execute `list(grad_tensors)`?

mari-linhares replied:
Probably adding these changes to hook should do the job, and in my opinion, is not that hacky.

DanyEle replied:
Yeah, hooking these changes should probably be sufficient. I just didn't have much of a clue as to where they should be placed.

DanyEle replied:
The error I was getting is the following one, BTW:

![image](https://user-images.githubusercontent.com/4907418/61272661-ee3f1780-a7a7-11e9-87f9-44ebed405875.png)

DanyEle commented:
Also, Issue #2318 has finally been resolved and closed, with no need for AutogradTensors. Nevertheless, it may still be useful to keep this PR for serializing and deserializing AutogradTensors.

DanyEle commented:
I added two test cases for AutogradTensor. 

1) Sending and receiving an AutogradTensor remotely.

2) Training a simple NN model with AutogradTensors. However, there seem to be some issues in the forward propagation phase. Maybe @LaRiffle  may take a look at the test case `test_train_remote_autograd_tensor`

here is the error I get

![image](https://user-images.githubusercontent.com/4907418/61271928-26455b00-a7a6-11e9-9764-2559d68e4d20.png)

DanyEle commented:
Hi @mari-linhares , I managed to fix the test case not passing. Silly me, I was extracing the model's weights locally and then trying to perform a forward pass then no weights were even present in the model!

The training procedure I used is simply the same one as in the test case `test_encrypted_training_with_linear_model` in the file `test_autograd.py`

midokura-silvia commented on .spyproject/codestyle.ini:
```diff
@@ -0,0 +1,6 @@
+[codestyle]
```
the files in the .spyproject folder should not be part of this pull request.

midokura-silvia commented on test/torch/tensors/test_autograd.py:
```diff
@@ -511,3 +511,63 @@ def test_get_float_prec_on_autograd_tensor(workers):
+    # let's check if the local version and the remote version have the same weigth
+    equal_weights = torch.all(
+        torch.eq(
+            model_local_trained.weight.copy().get().data, model_remote.weight.copy().get().data
```
why do you need to copy the weights before calling get()?

DanyEle replied:
There is no explicit reason here, but usually calling .copy().get() allows the weights to remain in the remote model for further operations, whereas .get() would just get them locally and remove them from the remote model.

midokura-silvia commented on test/torch/tensors/test_autograd.py:
```diff
@@ -511,3 +511,63 @@ def test_get_float_prec_on_autograd_tensor(workers):
+
+    assert equal_weights == 1
+
+    assert loss_remote.copy().get().item() < 500 and loss_local.item() < 500
```
Let's see how we can improve this second assert statement.
A loss < 500 seems very lax to me. Does the fitting work on this dataset work with 10 epochs? I guess it does, given that the classifier just needs to look at the first coordinate (linearly separable problem, much easier than xor). 
If the training is stable, we could ask for both remote and local loss to decrease through training. 
I just saw that you send the local model after training to the remote location to train again. That makes it a bit more tricky to check the loss. Any propositions?

DanyEle replied:
Yeah, it is indeed a bit more tricky because it looks like the local model is being affected too, following the remote training. I tried to detach all the parameters of the local model, but that didn't make a difference.

As you suggested, we could probably check for a monotonic decrease of the loss during the training procedure both for the local and the remote training procedure.

mari-linhares replied:
Agree.

midokura-silvia commented on test/torch/tensors/test_autograd.py:
```diff
@@ -511,3 +511,63 @@ def test_get_float_prec_on_autograd_tensor(workers):
+            opt.step()
+        return (loss, model_input)
+
+    # Training procedure to train an input model, be it remote or local
```
The comment ( # Training procedure) should be attached to the nested train() function

DanyEle replied:
Yeah, I'll fix it.

midokura-silvia commented on test/torch/tensors/test_autograd.py:
```diff
@@ -511,3 +511,63 @@ def test_get_float_prec_on_autograd_tensor(workers):
+    model_local = nn.Linear(2, 1)
+
+    # Local training
+    loss_local, model_local_trained = train(model_local, data_local, target_local)
```
model_local and model_local_trained point to the same model, right?

DanyEle replied:
Yep, they actually point to the same model, so we could possibly get rid of the model_local_trained variable.

midokura-silvia replied:
If you calculate the loss before training you could see a decrease. But probably not for the second training round (the remote one), as there are only 4 data points and you already trained the 3 parameters of the model for 10 steps.

DanyEle commented:
Should be good to go now! I added the checks for loss decrease both for the remote and the local version.

mari-linhares commented on syft/frameworks/torch/tensors/interpreters/multi_pointer.py:
```diff
@@ -103,7 +103,6 @@ def shape(self) -> torch.Size:
     def dim(self) -> int:
         """This method fixes the error that the result of dim was a list of ints
         stored inside a multipointer tensor"""
-
```
Can you undo this change? Just so we don't track in unedeed file.

DanyEle replied:
Should be done now

mari-linhares commented on test/torch/tensors/test_autograd.py:
```diff
@@ -511,3 +511,79 @@ def test_get_float_prec_on_autograd_tensor(workers):
+    remote_tensor = random_tensor.send(alice, local_autograd=True)
+    local_tensor = remote_tensor.get()
+    # check if the tensor sent is equal to the tensor got from the remote version
+    equal_tensors = torch.all(torch.eq(local_tensor, random_tensor))
```
You can just have a boolean assertion here: `assert torch.all(...)` instead of asserting  that is equal to 1.

DanyEle replied:
Done!

mari-linhares commented on test/torch/tensors/test_autograd.py:
```diff
@@ -511,3 +511,79 @@ def test_get_float_prec_on_autograd_tensor(workers):
+    loss_remote, model_remote_trained = train(model_remote, data_remote, target_remote, remote=True)
+
+    # let's check if the local version and the remote version have the same weigth
+    equal_weights = torch.all(
```
Same comment about boolean assertion

midokura-silvia commented on syft/frameworks/torch/tensors/interpreters/autograd.py:
```diff
@@ -265,3 +265,60 @@ def float_precision(self):
+        chain = None
+        # send the child over
+        if hasattr(tensor, "child"):
+            chain = syft.serde._simplify(tensor.child)
```
Could be contracted into a ternary operator:
chain = syft.serde._simplify(tensor.child) if hasattr(tensor, "child") else None

DanyEle replied:
Done!

DanyEle commented:
Should be ready to be merged?

LaRiffle commented on test/torch/tensors/test_autograd.py:
```diff
@@ -541,3 +541,73 @@ def test_get_float_prec_on_autograd_tensor(workers):
+    alice = workers["alice"]
+    random_tensor = torch.randn(5, 3)
+    remote_tensor = random_tensor.send(alice, local_autograd=True)
+    local_tensor = remote_tensor.get()
```
This tests doesn't work: there are no autogradtensor created here because:
- you need to set requires_grad to true on tensors
- send(..., local_autograd=True) creates an autogradtensor locally, you would need to send this object to another worker to have the expected behaviour

I would suggest check types of your local and remote objects in the tests to avoid this and to improve readability :)

DanyEle replied:
Alright, so would it be fine  to have:

`random_tensor = torch.randn(5, 3)` ?

LaRiffle replied:
Better to not use randn I think to be more elegant, try `tensor = torch.tensor([[3.0, 2], [-1, 2]], requires_grad=True)` for example

LaRiffle commented on test/torch/tensors/test_autograd.py:
```diff
@@ -541,3 +541,73 @@ def test_get_float_prec_on_autograd_tensor(workers):
+
+    # Some Toy Data
+    data_local = torch.tensor([[0, 0], [0, 1], [1, 0], [1, 1.0]])
+    target_local = torch.tensor([[0], [0], [1], [1.0]])
```
Same here!

DanyEle replied:
So, I need to add a requires_grad=True on these tensors?

LaRiffle replied:
try `data_local = torch.tensor([[0, 0], [0, 1], [1, 0], [1, 1.0]], requires_grad=True) ` instead

LaRiffle commented on test/torch/tensors/test_autograd.py:
```diff
@@ -541,3 +541,73 @@ def test_get_float_prec_on_autograd_tensor(workers):
+    # Remote training, setting autograd for the data and the targets
+    data_remote = data_local.send(worker, local_autograd=True)
+    target_remote = target_local.send(worker, local_autograd=True)
+    model_remote = model_local.send(worker, local_autograd=True)
```
So there is a bug here: Autograd should be created because model parameters have require_grad to True but it's not currently working :/

iamtrask commented:
Am I correct in reading that when you move a set of AutogradTensors to another machine, that they actually create the graph on that machine?

iamtrask commented:
Also, it doesn't seem like the gradient gets serialized if it exists. Is this true?

DanyEle commented:
> Also, it doesn't seem like the gradient gets serialized if it exists. Is this true?

I confirm this. This doesn't just happen for Autograd tensors though, but also for remote tensors being transmitted with PySyft.

DanyEle commented:
I implemented the fixes suggested by @LaRiffle. However, when trying to send a tensor with a `requires_grad=True`, is being produced. I thought this error was due to the fact that I was not serializing/deserializing the requires_grad attribute, so I went ahead and added serialization and de-serialization for that attribute. Nevertheless, the error is still there in the forward propagation phase into the model.

![image](https://user-images.githubusercontent.com/4907418/62272335-264a8980-b43b-11e9-95b6-63bd824975ac.png)
![image](https://user-images.githubusercontent.com/4907418/62272357-35313c00-b43b-11e9-9455-6dc29ff76436.png)
